rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,248 @@
+> coordination, Hudi is able to offer a few flexible deployment models that offer greater concurrency over pure OCC 
+> approaches that just track table snapshots."
+  
+In the multi-writer scenario, Hudi's existing conflict detection occurs after the writer finished writing the data 
+and before committing the metadata. In other words, the writer will only detect the occurrence of the conflict
+when it starts to commit, although all calculations and data writing have been completed, which will cause a waste 
+of resources.
+
+For example:
+
+Now there are two writing jobs: job1 will write 10M data to the Hudi table, including update file group 1. 
+Another job2 will write 100G to the Hudi table, and will also update the same file group 1. 
+
+Job1 finished and committed to Hudi successfully. After a few hours, job2 finished writing data files(100G) and start 
+to commit metadata. At this time, a conflict compared with job1 was found, and the job2 had to be aborted and re-run 
+after failure. Obviously, a lot of computing resources and time are wasted for job2.
+
+
+Hudi currently has two important mechanisms, marker mechanism and heartbeat mechanism:
+1. Marker mechanism can track all the files that are part of an active write.
+2. Heartbeat mechanism that can track all active writers to a Hudi table.
+
+
+Based on marker and heartbeat, this RFC design a new conflict detection: Early Conflict Detection.
+Before the writer creates the marker and before it starts to write the file, Hudi will perform this new conflict
+detection, trying to detect the writing conflict directly or get the async conflict check result(Timeline-Based) as
+early as possible and abort the writer when the conflict occurs, so that we can release resources as soon as possible
+and improve resource utilization.
+
+
+## Implementation
+Here is the high level workflow of early conflict detection as figure1 showed. 
+As we can see, only both `supportsOptimisticConcurrencyControl` and `isEarlyConflictDetectionEnable` are true, we 
+could use this early conflict detection feature. Else, we will skip this check and create marker directly.
+
+![](figure1.png)
+
+The three important steps marked in red in Figure 1 are introduced one by one as follows:
+
+### [1] Check Marker Conflict
+
+As we know, we have two ways to create and maintain markers:
+1. DirectWriteMarkers: individual marker file corresponding to each data file is directly created by the writer.
+2. TimelineServerBasedWriteMarkers: marker operations are all handled at the timeline service which serves as a proxy
+
+Therefore, for different types of Marker we have, we must implement the corresponding check marker conflict logic. 
+Here we expand the existing `ConflictResolutionStrategy` interface to ensure the scalability of checking marker conflict.
+
+![](flow1.png)
+
+In this design, we provide `DirectMarkerWithTransactionConflictResolutionStrategy` and 
```

Comment:
I do not see this class `DirectMarkerWithTransactionConflictResolutionStrategy` in flow1.png diagram. Please check it once.

Reply:
added.